internal/credsretriever/refreshing_cache.go

```diff
@@ -188,12 +196,18 @@ func (r *cachedCredentialRetriever) GetIamCredentials(ctx context.Context,
 		}
 	}
 
-	r.internalActiveRequestCache.Add(request.ServiceAccountToken, true)
-	defer r.internalActiveRequestCache.Delete(request.ServiceAccountToken)
+	r.internalActiveRequestCache.Add(request.ServiceAccountToken, nil)
 
 	log.WithField("cache-hit", 0).Tracef("Could not find entry in cache, requesting creds from delegate")
 
 	iamCredentials, metadata, err := r.callDelegateAndCache(ctx, request)
+	defer func() {
+		if err == nil {
+			r.internalActiveRequestCache.Delete(request.ServiceAccountToken)
+		} else {
+			r.internalActiveRequestCache.ReplaceWithExpire(request.ServiceAccountToken, err, defaultActiveRequestInterval)
+		}
+	}()
```

Comment:
non blocking nit: for a deferred func, you probably want to explicitly pass the error to the function


```suggestion
	defer func(err error) {
		if err == nil {
			r.internalActiveRequestCache.Delete(request.ServiceAccountToken)
		} else {
			r.internalActiveRequestCache.ReplaceWithExpire(request.ServiceAccountToken, err, defaultActiveRequestInterval)
		}
	}(err)
```

Reply:
Hi Micah, will follow up on these 2, I merged the change because this is actually covered by unit test, we are able to verify that error is stored in the memory cache.